src/Simplex/Chat/Library/Internal.hs

```diff
@@ -1024,65 +1029,134 @@ introduceToModerators :: VersionRangeChat -> User -> GroupInfo -> GroupMember ->
 introduceToModerators vr user gInfo@GroupInfo {groupId} m@GroupMember {memberRole, memberId} = do
   forM_ (memberConn m) $ \mConn -> do
     let msg =
-          if (maxVersion (memberChatVRange m) >= groupKnockingVersion)
+          if maxVersion (memberChatVRange m) >= groupKnockingVersion
             then XGrpLinkAcpt GAPendingReview memberRole memberId
             else XMsgNew $ MCSimple $ extMsgContent (MCText pendingReviewMessage) Nothing
     void $ sendDirectMemberMessage mConn msg groupId
   modMs <- withStore' $ \db -> getGroupModerators db vr user gInfo
-  let rcpModMs = filter (\mem -> memberCurrent mem && maxVersion (memberChatVRange mem) >= groupKnockingVersion) modMs
+  let rcpModMs = filter introduceMemP modMs
   introduceMember vr user gInfo m rcpModMs (Just $ MSMember $ memberId' m)
+  where
+    introduceMemP :: GroupMember -> Bool
+    introduceMemP mem =
+      memberCurrent mem
+        && groupMemberId' mem /= groupMemberId' m
+        && maxVersion (memberChatVRange mem) >= groupKnockingVersion
 
 introduceToAll :: VersionRangeChat -> User -> GroupInfo -> GroupMember -> CM ()
-introduceToAll vr user gInfo m = do
-  members <- withStore' $ \db -> getGroupMembersForIntroduction db vr user gInfo m
-  let recipients = filter memberCurrent members
+introduceToAll vr user gInfo m@GroupMember {relationsVector} = do
+  members <- withStore' $ \db -> getGroupMembers db vr user gInfo
+  let recipients = filter introduceMemP members
   introduceMember vr user gInfo m recipients Nothing
+  where
+    introduceMemP :: GroupMember -> Bool
+    introduceMemP mem = case relationsVector of
+      Nothing ->
+        memberCurrent mem && groupMemberId' mem /= groupMemberId' m
+      Just vec ->
+        let introducedMemsIdxs = getMemberRelationsIndexes (\r -> r == MRIntroduced || r == MRIntroducedTo || r == MRConnected) vec
```

Comment:
this should be made into a set, but it may be better to do direct lookup - I don't understand wha't the benefit of making it a list first, when direct lookup is more efficient. Why is array of indices needed here? and not lookup each member index directly while filtering? Seems unnecessarily complex and indirect.